File: app/core/safety.py
#!/usr/bin/env python3
"""
Multi-Layer Safety Engine.

Architecture — three layers checked in order:
  Layer 1 — Rule-based blocklist (fast, deterministic)
  Layer 2 — Intent Risk Classifier (heuristic scoring)
  Layer 3 — Model-based safety check (optional LLM judge)

Action Guard:
  if risk == HIGH or CRITICAL → block execution
  if risk == MEDIUM and strict_mode → block execution

Sandbox note:
  Actual sandboxed execution is delegated to the Execution Engine;
  this module only decides *whether* to allow it.
"""

from __future__ import annotations

import re
from typing import Callable, Coroutine, Optional

from app.config import settings
from app.models.schemas import ExtractedIntent, RiskLevel, SafetyVerdict

# ───────────── Layer 1: Blocked keywords / patterns ──────────────────────

BLOCKED_KEYWORDS = [
    "delete_all_data",
    "drop_database",
    "drop_table",
    "system_shutdown",
    "format_disk",
    "rm -rf",
    "truncate",
    "shutdown",
    "reboot",
    "erase_all",
    "destroy",
]

BLOCKED_PATTERNS = [
    re.compile(r"\bdrop\s+(database|table|schema)\b", re.IGNORECASE),
    re.compile(r"\bdelete\s+from\s+\*\b", re.IGNORECASE),
    re.compile(r"\btruncate\s+table\b", re.IGNORECASE),
    re.compile(r"\bformat\s+[a-z]:\b", re.IGNORECASE),
    re.compile(r"\brm\s+(-rf|--force)\b", re.IGNORECASE),
]

# ───────────── Layer 2: Intent risk map ──────────────────────────────────

_INTENT_RISK_MAP = {
    "delete_data": RiskLevel.HIGH,
    "drop_database": RiskLevel.CRITICAL,
    "system_shutdown": RiskLevel.CRITICAL,
    "send_message": RiskLevel.MEDIUM,
    "convert_format": RiskLevel.LOW,
    "summarize": RiskLevel.LOW,
    "generate_report": RiskLevel.LOW,
    "query_data": RiskLevel.LOW,
    "explain": RiskLevel.LOW,
    "unknown": RiskLevel.MEDIUM,
}

# ───────────── Public API ────────────────────────────────────────────────


async def check_safety(
    intent: ExtractedIntent,
    raw_text: str = "",
    *,
    llm_safety_call: Optional[
        Callable[[str], Coroutine[None, None, str]]
    ] = None,
) -> SafetyVerdict:
    """Run the full multi-layer safety pipeline.

    Parameters
    ----------
    intent:
        The extracted intent to evaluate.
    raw_text:
        Original user input (used by Layer 1 & 3).
    llm_safety_call:
        Optional async LLM callable for Layer 3 model-based check.
    """
    # ── Layer 1: Rule-based blocklist ────────────────────────────────
    verdict = _check_blocklist(raw_text, intent)
    if verdict is not None:
        return verdict

    # ── Layer 2: Intent risk classification ──────────────────────────
    risk = _classify_risk(intent)

    if risk in (RiskLevel.HIGH, RiskLevel.CRITICAL):
        return SafetyVerdict(
            allowed=False,
            risk_level=risk,
            reason=f"Intent '{intent.intent}' classified as {risk.value} risk.",
            blocked_by="intent_risk",
        )

    if risk == RiskLevel.MEDIUM and settings.safety_strict_mode:
        return SafetyVerdict(
            allowed=False,
            risk_level=risk,
            reason=f"Intent '{intent.intent}' is medium risk; strict mode is enabled.",
            blocked_by="intent_risk",
        )

    # ── Layer 3: Optional model-based safety check ───────────────────
    if llm_safety_call is not None:
        model_verdict = await _model_safety_check(
            intent, raw_text, llm_safety_call
        )
        if model_verdict is not None:
            return model_verdict

    # ── All clear ────────────────────────────────────────────────────
    return SafetyVerdict(
        allowed=True,
        risk_level=risk,
        reason="All safety checks passed.",
    )


# ─────────────────── Internal helpers ────────────────────────────────────


def _check_blocklist(
    raw_text: str, intent: ExtractedIntent
) -> Optional[SafetyVerdict]:
    """Layer 1 — fast keyword and regex check."""
    combined = f"{raw_text} {intent.intent}".lower()

    for kw in BLOCKED_KEYWORDS:
        if kw in combined:
            return SafetyVerdict(
                allowed=False,
                risk_level=RiskLevel.CRITICAL,
                reason=f"Blocked keyword detected: '{kw}'",
                blocked_by="rule",
            )

    for pattern in BLOCKED_PATTERNS:
        if pattern.search(combined):
            return SafetyVerdict(
                allowed=False,
                risk_level=RiskLevel.CRITICAL,
                reason=f"Blocked pattern matched: {pattern.pattern}",
                blocked_by="rule",
            )

    return None  # no match — continue to next layer


def _classify_risk(intent: ExtractedIntent) -> RiskLevel:
    """Layer 2 — map intent to risk level."""
    return _INTENT_RISK_MAP.get(intent.intent, RiskLevel.MEDIUM)


async def _model_safety_check(
    intent: ExtractedIntent,
    raw_text: str,
    llm_call,
) -> Optional[SafetyVerdict]:
    """Layer 3 — ask LLM to judge safety (optional)."""
    prompt = (
        "You are a safety classifier. Given the following user request and "
        "extracted intent, respond with ONLY a JSON object:\n"
        '{"safe": true/false, "reason": "..."}\n\n'
        f"User request: {raw_text}\n"
        f"Intent: {intent.intent}\n"
        f"Parameters: {intent.parameters}\n"
    )
    try:
        raw = await llm_call(prompt)
        import json
        result = json.loads(raw.strip())
        if not result.get("safe", True):
            return SafetyVerdict(
                allowed=False,
                risk_level=RiskLevel.HIGH,
                reason=result.get("reason", "Model-based check flagged this action."),
                blocked_by="model",
            )
    except Exception:
        pass  # if model call fails, default to allowing (layers 1+2 already passed)

    return None

File: npm/src/core/safety.ts
/**
 * Kavach — 3-Layer Safety Engine.
 *
 * Layer 1: Rule-based blocklist (keywords + regex) — zero tokens, < 1ms
 * Layer 2: Intent risk classification (deterministic map)
 * Layer 3: Optional LLM judge (async, costs tokens)
 *
 * Action Guard:
 *   CRITICAL / HIGH → always block
 *   MEDIUM + strict mode → block
 *   LOW → always allow
 */

import { RiskLevel } from "../types.js";
import type { ExtractedIntent, SafetyVerdict, LLMCall } from "../types.js";

// ─── Configuration ──────────────────────────────────────────────────────

let strictMode = false;

export function setStrictMode(enabled: boolean): void {
  strictMode = enabled;
}

// ─── Layer 1: Blocklist ─────────────────────────────────────────────────

const BLOCKED_KEYWORDS: string[] = [
  "delete_all_data",
  "drop_database",
  "drop_table",
  "system_shutdown",
  "format_disk",
  "rm -rf",
  "truncate",
  "shutdown",
  "reboot",
  "erase_all",
  "destroy",
];

const BLOCKED_PATTERNS: RegExp[] = [
  /\bdrop\s+(database|table|schema)\b/i,
  /\bdelete\s+from\s+\*/i,
  /\btruncate\s+table\b/i,
  /\bformat\s+[a-z]:/i,
  /\brm\s+(-rf|--force)\b/i,
];

function checkBlocklist(rawText: string, intent: ExtractedIntent): SafetyVerdict | null {
  const combined = `${rawText} ${intent.intent}`.toLowerCase();

  for (const kw of BLOCKED_KEYWORDS) {
    if (combined.includes(kw)) {
      return {
        allowed: false,
        riskLevel: RiskLevel.CRITICAL,
        reason: `Blocked keyword detected: '${kw}'`,
        blockedBy: "rule",
      };
    }
  }

  for (const pattern of BLOCKED_PATTERNS) {
    if (pattern.test(combined)) {
      return {
        allowed: false,
        riskLevel: RiskLevel.CRITICAL,
        reason: `Blocked pattern matched: ${pattern.source}`,
        blockedBy: "rule",
      };
    }
  }

  return null; // no match
}

// ─── Layer 2: Intent risk map ───────────────────────────────────────────

const INTENT_RISK_MAP: Record<string, RiskLevel> = {
  delete_data: RiskLevel.HIGH,
  drop_database: RiskLevel.CRITICAL,
  system_shutdown: RiskLevel.CRITICAL,
  send_message: RiskLevel.MEDIUM,
  convert_format: RiskLevel.LOW,
  summarize: RiskLevel.LOW,
  generate_report: RiskLevel.LOW,
  query_data: RiskLevel.LOW,
  explain: RiskLevel.LOW,
  unknown: RiskLevel.MEDIUM,
};

function classifyRisk(intent: ExtractedIntent): RiskLevel {
  return INTENT_RISK_MAP[intent.intent] ?? RiskLevel.MEDIUM;
}

// ─── Layer 3: Optional LLM judge ────────────────────────────────────────

async function modelSafetyCheck(
  intent: ExtractedIntent,
  rawText: string,
  llmCall: LLMCall
): Promise<SafetyVerdict | null> {
  const prompt =
    `You are a safety classifier. Given the following user request and ` +
    `extracted intent, respond with ONLY a JSON object:\n` +
    `{"safe": true/false, "reason": "..."}\n\n` +
    `User request: ${rawText}\n` +
    `Intent: ${intent.intent}\n` +
    `Parameters: ${JSON.stringify(intent.parameters)}\n`;

  try {
    const raw = await llmCall(prompt);
    const result = JSON.parse(raw.trim()) as { safe?: boolean; reason?: string };
    if (!result.safe) {
      return {
        allowed: false,
        riskLevel: RiskLevel.HIGH,
        reason: result.reason ?? "Model-based check flagged this action.",
        blockedBy: "model",
      };
    }
  } catch {
    // If model fails, default to allowing (layers 1+2 already passed)
  }

  return null;
}

// ─── Public API ─────────────────────────────────────────────────────────

/**
 * Run the full 3-layer safety pipeline.
 */
export async function checkSafety(
  intent: ExtractedIntent,
  rawText: string = "",
  llmCall?: LLMCall
): Promise<SafetyVerdict> {
  // Layer 1: Blocklist
  const blocklistVerdict = checkBlocklist(rawText, intent);
  if (blocklistVerdict) return blocklistVerdict;

  // Layer 2: Risk classification
  const risk = classifyRisk(intent);

  if (risk === RiskLevel.HIGH || risk === RiskLevel.CRITICAL) {
    return {
      allowed: false,
      riskLevel: risk,
      reason: `Intent '${intent.intent}' classified as ${risk} risk.`,
      blockedBy: "intent_risk",
    };
  }

  if (risk === RiskLevel.MEDIUM && strictMode) {
    return {
      allowed: false,
      riskLevel: risk,
      reason: `Intent '${intent.intent}' is medium risk; strict mode is enabled.`,
      blockedBy: "intent_risk",
    };
  }

  // Layer 3: Optional model check
  if (llmCall) {
    const modelVerdict = await modelSafetyCheck(intent, rawText, llmCall);
    if (modelVerdict) return modelVerdict;
  }

  // All clear
  return {
    allowed: true,
    riskLevel: risk,
    reason: "All safety checks passed.",
  };
}

// Re-export helpers for testing
export { checkBlocklist as _checkBlocklist, classifyRisk as _classifyRisk };
